src/Search/LexiconfreeBeamSearch/LexiconfreeBeamSearch.cc

```diff
@@ -228,6 +231,57 @@ std::string LexiconfreeBeamSearch::LabelHypothesis::toString() const {
     return ss.str();
 }
 
+void LexiconfreeBeamSearch::tokenPruning(std::vector<size_t>& indices, std::vector<Score> extensionScores, size_t numUnfinishedHyps) {
+    #pragma omp parallel for
+        for (size_t hypIndex = 0ul; hypIndex < numUnfinishedHyps; ++hypIndex) {
+            // Start and end index of scores coming from current hypothesis
+            size_t start = hypIndex * lexicon_->nLemmas();
+            size_t end   = start + lexicon_->nLemmas();
+
+            // Make sure the best topKTokens_ scores are moved to the beginning of the [start, end) interval
+            std::nth_element(indices.begin() + start, indices.begin() + start + topKTokens_, indices.begin() + end,
+                             [&extensionScores](size_t a, size_t b) {
+                                 return extensionScores[a] < extensionScores[b];
+                             });
+
+            // Copy the topKTokens_ best indices to the front of indices (note: this location doesn't overlap with [start, end) of the next hyps because topKTokens_ < nLemmas)
+            if (hypIndex > 0ul) {
+                std::copy(indices.begin() + start, indices.begin() + start + topKTokens_, indices.begin() + hypIndex * topKTokens_);
+            }
+        }
+
+    indices.resize(numUnfinishedHyps * topKTokens_);  // Throw away all pruned indices
+}
+
+template <typename Hypotheses, typename CompareFunc>
+void LexiconfreeBeamSearch::beamPruning(Hypotheses& hypotheses, CompareFunc compare) {
```

Comment:
```suggestion
template <typename T>
void LexiconfreeBeamSearch::beamPruning(std::vector<T>& hypotheses, std::function<bool,(T const&, T const&)>&& compare) {
```
You can rewrite the template definition to make it clearer how `compare` is supposed to look like.